apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts

```diff
@@ -305,6 +319,222 @@ export class TypeScriptBuilder extends SubprocessRunnerBase<ITypeScriptBuilderCo
+      __innerEmit?: TTypescript.Program['emit'];
+    }
+
+    const program: IProgramWithMultiEmit = innerProgram;
+
+    let { __innerEmit: innerEmit, __innerGetCompilerOptions: innerGetCompilerOptions } = program;
+
+    if (!innerGetCompilerOptions) {
+      program.__innerGetCompilerOptions = innerGetCompilerOptions = program.getCompilerOptions;
+    }
+
+    if (!innerEmit) {
+      program.__innerEmit = innerEmit = program.emit;
+    }
+
+    interface IKindOptions {
+      compilerOptions: TTypescript.CompilerOptions;
+    }
+
+    let foundPrimary: boolean = false;
+    let defaultModuleKind: TTypescript.ModuleKind;
+
+    const multiEmitMap: Map<ICachedEmitModuleKind, IKindOptions> = new Map();
+    for (const moduleKindToEmit of this._moduleKindsToEmit) {
+      const kindOptions: IKindOptions = {
+        compilerOptions: moduleKindToEmit.isPrimary
+          ? {
+              ...innerGetCompilerOptions()
+            }
+          : {
+              ...innerGetCompilerOptions(),
+              module: moduleKindToEmit.moduleKind,
+              outDir: moduleKindToEmit.outFolderPath,
+
+              // Don't emit declarations for secondary module kinds
+              declaration: false,
+              declarationMap: false
+            }
+      };
+      if (!kindOptions.compilerOptions.outDir) {
+        throw new InternalError('Expected compilerOptions.outDir to be assigned');
+      }
+      multiEmitMap.set(moduleKindToEmit, kindOptions);
+
+      if (moduleKindToEmit.isPrimary) {
+        if (foundPrimary) {
+          throw new Error('Multiple primary module emit kinds encountered.');
+        } else {
+          foundPrimary = true;
+        }
+
+        defaultModuleKind = moduleKindToEmit.moduleKind;
+      }
+    }
+
+    const changedFiles: Set<IExtendedSourceFile> = new Set();
+
+    program.emit = (
+      targetSourceFile?: TTypescript.SourceFile,
+      writeFile?: TTypescript.WriteFileCallback,
+      cancellationToken?: TTypescript.CancellationToken,
+      emitOnlyDtsFiles?: boolean,
+      customTransformers?: TTypescript.CustomTransformers
+    ) => {
+      if (emitOnlyDtsFiles) {
+        return program.__innerEmit!(
+          targetSourceFile,
+          writeFile,
+          cancellationToken,
+          emitOnlyDtsFiles,
+          customTransformers
+        );
+      }
+
+      if (targetSourceFile && changedFiles) {
+        changedFiles.add(targetSourceFile as IExtendedSourceFile);
+      }
+
+      const originalCompilerOptions: TTypescript.CompilerOptions = program.__innerGetCompilerOptions!();
+
+      let defaultModuleKindResult: TTypescript.EmitResult;
+      const diagnostics: TTypescript.Diagnostic[] = [];
+      let emitSkipped: boolean = false;
+      try {
+        for (const [moduleKindToEmit, kindOptions] of multiEmitMap) {
+          const kindCompilerOptions: TTypescript.CompilerOptions = kindOptions.compilerOptions;
```

Comment:
The wrapper object here can go away; it was a temporary artifact of me trying to cache the wrapped writeFile call, until I realized that doing so isn't safe to do so, since the function passed in is unique per call when loaded from a BuilderProgram.